### PR TITLE
Cancel background work 

### PR DESCRIPTION
**Pull request by dsyme**
This is a WIP PR to extend #3061.

In #3061 I noticed that therre we some long running, non-cancellable operations on the background thread.  This means, for example, that if you 
1. start by opening a file in a project P, 
2. switch to a file or script unrelated to project P

then there will be no intellisense or colorization in the script until the basic preparation of project P (and possibly the complete checking of its dependencies) is finished

This PR cancels background work when new foreground work arrives that will implicitly start replacement background work (e.g. to prepare a script).

### CODE REVIEW

File: src/fsharp/vs/Reactor.fs [hunk 1]
```diff
@@ -81,33 +83,49 @@ type Reactor() =
...
+                            Trace.TraceWarning("FCS: {0}<-- {1}.{2}, took {3} ms ({4})", msg, userOpName, opName, span.TotalMilliseconds, opArg)
                         return! loop (bgOpOpt, onComplete, false)
                     | Some (WaitForBackgroundOpCompletion channel) -> 
                         match bgOpOpt with 
                         | None -> ()
                         | Some (bgUserOpName, bgOpName, bgOpArg, bgOp) -> 
-                            Trace.TraceInformation("Reactor: --> wait for background {0}.{1} ({2}), remaining {3}", bgUserOpName, bgOpName, bgOpArg, inbox.CurrentQueueLength)
-                            while bgOp ctok do 
+                            Trace.TraceInformation("FCS: --> wait for background {0}.{1} ({2}), remaining {3}", bgUserOpName, bgOpName, bgOpArg, inbox.CurrentQueueLength)
+                            bgOpCts <- new CancellationTokenSource()
```
Comment (vasily-kirichenko) [inline src/fsharp/vs/Reactor.fs]: Shouldn't we dispose the old bgOpCts first?

File: src/fsharp/vs/Reactor.fs [hunk 2]
```diff
@@ -81,33 +83,49 @@ type Reactor() =
...
                     | None -> 
                         match bgOpOpt, onComplete with 
                         | _, Some onComplete -> onComplete.Reply()
                         | Some  (bgUserOpName, bgOpName, bgOpArg, bgOp), None -> 
-                            Trace.TraceInformation("Reactor: --> background step {0}.{1} ({2})", bgUserOpName, bgOpName, bgOpArg)
+                            Trace.TraceInformation("FCS: --> background step {0}.{1} ({2})", bgUserOpName, bgOpName, bgOpArg)
                             let time = Stopwatch()
                             time.Start()
-                            let res = bgOp ctok
+                            bgOpCts <- new CancellationTokenSource()
```
Comment (vasily-kirichenko) [inline src/fsharp/vs/Reactor.fs]: Same as above.

File: vsintegration/src/FSharp.Editor/Options/EditorOptions.fs [hunk 3]
```diff
@@ -10,7 +10,7 @@ open OptionsUIHelpers
 
 module DefaultTuning = 
     let SemanticColorizationInitialDelay = 0 (* milliseconds *)
-    let UnusedDeclarationsAnalyzerInitialDelay = 1000 (* milliseconds *)
+    let UnusedDeclarationsAnalyzerInitialDelay = 0 (* milliseconds *)
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Options/EditorOptions.fs]: 👍 :)

Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Options/EditorOptions.fs]: Yes I think this is ok.

File: src/fsharp/vs/Reactor.fs (hunk 1 above)
Comment (dsyme) [inline src/fsharp/vs/Reactor.fs]: Yes, thanks

**Comment by dsyme** [comment]
A quick update on this:  I've been trying to use this in practice to get a sense if the cancellation makes for a better overall experience - for example I'm vaguely concerned there might be situations where the background builder never really makes progress because so many requests are coming in to FCS from things like the "SimplifyName" code fix.

I'll keep using it myself and keep an eye on things.

**Comment by dsyme** [comment]
This is now ready

**Comment by dsyme** [comment]
For the record, @KevinRansom asked me why I was hesitant about this PR. My reply:

> It's not flakey (unless the act of cancellation exposes some latent bug). It cancels the slice of the background build if language service requests come in.  That could in theory mean that the background build never progresses if there is a never ending flood of language service requests (e.g. you endlessly type, move the mouse etc. very fast).  But that would have happened anyway prior to the PR. So it’s ok.

**Pull request merged**